tuf/ngclient/updater.py

```diff
@@ -122,12 +131,80 @@ def refresh(self) -> None:
             RepositoryError: Metadata failed to verify in some way
             DownloadError: Download of a metadata file failed in some way
         """
+        try:
+            version_bytes = self._load_local_metadata("spec_version")
+            self._spec_version = json.loads(version_bytes.decode("utf-8"))[
+                "version"
+            ]
+
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise
+            self._spec_version = "1"
+
+        repository_versions_and_paths = self._get_repository_versions()
+
+        repository_versions = [
+            i["version"] for i in repository_versions_and_paths
+        ]
+
+        # Updating self.spec_version
+        spec_version, message = _get_spec_version(
+            repository_versions, self._spec_version, self._supported_versions
+        )
+
+        if message:
+            logger.warning(message)
+        self._spec_version = spec_version
+
+        ordered_version_paths = []
 
-        self._load_root()
+        def filter_fn(a):
+            return a["version"] == version and version <= int(spec_version)
+
+        for version in sorted(repository_versions):
+            path = list(
+                filter(
+                    filter_fn,
+                    repository_versions_and_paths,
+                )
+            )
+            if len(path) > 0:
+                ordered_version_paths.append(path[0]["path"])
+
+        for i in repository_versions_and_paths:
+            if i["version"] == spec_version:
+                self._spec_version_dir = i["path"]
+
+        # persist the version number
+        self._persist_metadata(
+            "spec_version",
+            json.dumps({"version": spec_version}).encode("utf-8"),
+        )
+
+        self._load_root(ordered_version_paths)
```

Comment:
Thinking about this: it's totally possible that v2 root format will look totally different, so it seems wrong to try to load them the same way.

I've thought about implementing TAP-14 as a *wrapper* around the Updater. Basically, it would expose the same interface and keep a dict of `{MajorVersion: Updater}`.

Then, its job is to (1) parse the `supported-versions` and (2) delegate appropriately. This keeps the complexity of TAP-14 out of the Updater itself, and provides a more natural way to support a very-different v2.

Reply:
I agree in general, but `_load_root` is unique in that it has to look in directories for multiple supported versions in order to ensure that no root metadata is missed. If the root format is changed, then this function could be updated as part of the client update to use the correct format for the current root file that is being checked.

Reply:
> I've thought about implementing TAP-14 as a _wrapper_ around the Updater. Basically, it would expose the same interface and keep a dict of `{MajorVersion: Updater}`.

This is a great idea. Legacy python-tuf implemented [TAP 4: Multiple repository consensus on entrusted targets](https://github.com/theupdateframework/taps/blob/master/tap4.md) (mapping file) support with a wrapper around the updater (see [`MultiRepoUpdater`](https://github.com/theupdateframework/python-tuf/blob/febe6c3dc30195078a5487d68d7db708e7de8479/tuf/client/updater.py#L161)). It could be useful to experiment with multiple wrappers around the updater and how they would interact, so that we can determine whether we need to come up with another pattern so that TAP implementations aren't mutually exclusive.

Reply:
That's a great point. We're hoping to take that approach in go-tuf for TAP 4 as well!

Interactions sound a little tricky...I *think* we're okay, we'd put the TAP-4 wrapper as the outermost layer?

We can cross that bridge when we come to it.

Reply:
Yeah, my hunch is that ordering with TAP 4 at the outermost layer makes sense for TAP 4 + TAP 14.